stellargraph/data/edge_cache.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Provides a standarised way to access graph neighbours via edge data,
+possibly using optimised caching.
+"""
+
+__all__ = ["EdgeCache", "EdgeCacheType"]
+
+from enum import Enum
+from typing import Iterable, Any
+import itertools
+
+from .edge_data import EdgeData
+from .node_data import NodeData
+
+
+class EdgeCacheType(Enum):
+    # No caching of edges
+    NONE = 1
+    # Cache edges, ignoring types
+    UNTYPED = 2
+    # TYPED = 3
+
+
+class EdgeCacheFactory(type):
+    """
+    Private class for instantiating the edge cache interface from
+    user-supplied information.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        if cls is EdgeCache:
+            if len(args) > 0:
+                cache_type = args[0]
+                args = args[1:]
+            else:
+                cache_type = kwargs.pop("cache_type", None)
+            if cache_type is None:
+                # Supply default
+                cache_type = EdgeCacheType.UNTYPED
+            if not isinstance(cache_type, EdgeCacheType):
+                raise TypeError("Invalid cache type: {}".format(cache_type))
+            if cache_type == EdgeCacheType.NONE:
+                return NoEdgeCache(*args, **kwargs)
+            if cache_type == EdgeCacheType.UNTYPED:
+                return UntypedEdgeCache(*args, **kwargs)
+            raise ValueError("Unknown cache type: {}".format(cache_type))
+        else:
+            return type.__call__(cls, *args, **kwargs)
+
+
+class EdgeCache(metaclass=EdgeCacheFactory):
```

Comment:
The two specific data-types that @adocherty wanted to support are Pandas data-frames and a dictionary mapping types to blocks of data. My previous experiments also allowed NumPy arrays and lists of data, but I took these out for simplicity.

My basic goal was to have a unified interface, e.g. `NodeData` or `EdgeData`, that could efficiently reuse the data in the underlying data structure wherever possible, to avoid memory copying. The explosion of tests you mentioned is a valid concern, but in practice one only needs to test the (currently two) `NodeData` implementations and the (currently two) `EdgeData` implementations (some of that testing exists). I don't think it is useful beyond a basic test to combine the cross-product of all of these tests in `StellarGraph`.

Reply:
The `EdgeCache` was an add-on that was required to speed up neighbourhood queries. The Single Responsibility principle, coupled with some deep thinking, suggests that `EdgeData` should not have to support caching, so it was designed as a wrapper around `EdgeData`.

Reply:
In a very real sense, the `NodeData` implementation is exactly a mechanism that takes one of a variety of input data types and provides one output data type (of currently two). Previous experiments used an explicit method to do this, with explicitly documented arguments - I kept changing my mind about whether to expose this method or not.

Reply:
I agree I have probably gone overboard with the factory instantiation approach. The benefit in the case of `StandardStellarGraph` versus `NetworkXStellarGraph` lies precisely in the fact that they currently have two very different argument lists, and this problem is hidden from the user. I'm not sure if there is a cleaner way to handle polymorphic arguments.

Reply:
I'm going to have to keep thinking about the rest of your comments, but for now:

> The benefit in the case of StandardStellarGraph versus NetworkXStellarGraph lies precisely in the fact that they currently have two very different argument lists, and this problem is hidden from the user

I don't think the problem can be hidden from the user: the possible valid argument lists is somewhat fundamentally relevant to the user, since passing in a valid argument list is the only way to create a `StellarGraph` and use the rest of the library. If they can't work out how to construct this (i.e. what's valid to pass to a constructor of `StellarGraph`), they cannot use the library.

Reply:
True, but you're saying that means we must always have a well-defined argument list so IDEs can help us with parameters. That aim is helpful and desirable, but not even standard libraries like Keras always adhere to that.

My argument is not that we don't care or shouldn't aim for a consistent, explicit interface, but more that it is still too early to know what that looks like. For example, the `NetworkXStellarGraph` allows a `node_type_default` argument, but `StandardStellarGraph` calls it `default_node_type`. Must we explicitly allow all variants up front in the argument list, just because we haven't rationalised the nomenclature? I just want to defer that decision until we know more.